### PR TITLE
[Merged by Bors] - change how to select bevy-glsl-to-spirv or shaderc

### LINKED ISSUE #898: bevy fails to compile/run on musl based linux distros
**Bevy version**
0.3/master also fails the same way

**Operating system & version**
Alpine linux edge

**What you did**
Attempt to compile and run a bevy example on alpine linux.

**What you expected to happen**

the sample would run

**What actually happened**

The engine fails to compile when compiling cart/glsl-to-spirv

**Additional information**

The engine fails to compile as `x86_64-unknown-linux-gnu` is being used but alpine expects `x86_64-unknown-linux-musl`
If I edit the glsl-to-spirv build.rs files to use `-musl`the compilation succeeds however, it then fails to launch and exits with a segfault to the usage of the precompiled static libraries expecting glibc (as far as i can understand anyway).

Also worth mentioning, if I dont use the fast compile .cargo/config i get a huge error that ends with the log below.
```
  = note: /usr/lib/gcc/x86_64-alpine-linux-musl/10.2.1/../../../../x86_64-alpine-linux-musl/bin/ld: /home/alex/glsl-to-spirv/build/linux-musl/libglslang.glsltospirv.a(Constant.cpp.o): undefined reference to symbol 'acos'
          /usr/lib/gcc/x86_64-alpine-linux-musl/10.2.1/../../../../x86_64-alpine-linux-musl/bin/ld: /lib/libc.musl-x86_64.so.1: error adding symbols: DSO missing from command line
          collect2: error: ld returned 1 exit status
```

Since reporting this I've also tried building with shaderc adding conditionals for musl/linux which when running an example segfaults

### LINKED ISSUE #1078: Cannot compile Bevy with 32 bit toolchain -- i686-pc-windows-msvc
**Bevy version**

git: version = "0.3.0"
AND
Crates.io: version = "0.3.0"

**Operating system & version**

Windows 10 - 20H2

**What you did**

I have a project that works completely when using `stable-x86_64-pc-windows-msvc`. I tried changing it to `i686-pc-windows-msvc` and can no longer compile.

**What actually happened**

cargo build --target=i686-pc-windows-msvc
   Compiling bevy-glsl-to-spirv v0.2.1
   Compiling serde_json v1.0.60
error: failed to run custom build command for `bevy-glsl-to-spirv v0.2.1`

Caused by:
  process didn't exit successfully: `build\bevy-glsl-to-spirv-84bcc83f6d89e5d2\build-script-build` (exit code: 101)
  --- stderr
  thread 'main' panicked at 'Missing target support i686-pc-windows-msvc', .cargo\registry\src\github.com-1ecc6299db9ec823\bevy-glsl-to-spirv-0.2.1\build/build.rs:24:9
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
warning: build failed, waiting for other jobs to finish...
error: build failed

**Additional information**

Additionally, after seeing this failure, I created a blank project and included Bevy as a dependency. It too failed to run.

### LINKED ISSUE #1348: Use `bevy-glsl-to-spirv` only on the supported targets, default to `shaderc` otherwise.
The current configuration uses `bevy-glsl-to-spirv` by default, while selectively enabling `shaderc` for specific targets.

I think this is the wrong way around.

`bevy-glsl-to-spirv` provides precompiled binaries to help with builds on the officially supported targets. By design, it can only work for specifically-supported build targets.

`shaderc` has additional build-time requirements and complexity, but it supports a much wider range of compilation targets.

It doesn't make sense to me that `bevy-glsl-to-spirv` is the blanket default, and `shaderc` is on a special whitelist.

This results in compilation failures whenever anyone tries building bevy on a new target (`aarch64-unknown-linux-gnu`? `x86_64-unknown-linux-musl`? etc.), leading to issues like #898. I've seen many instances of similar reports; someone tries to compile bevy on their raspberry pi or whatever, only to hit the `bevy-glsl-to-spirv` error.

With `shaderc`, such platforms should be able to work (and would have likely worked out of the box for users attempting them for the first time).

Can we flip the configuration around? Make `bevy-glsl-to-spirv` conditionally-enabled on the platforms it supports, and default to `shaderc` otherwise?

This would keep builds exactly the same for any existing users of bevy, but it will help make bevy work on new targets more seamlessly.

### LINKED ISSUE #1942: Bevy 0.4 -> 0.5 broke aarch64-unknown-linux-gnu
On `armv8.2a`, 0.5 fails on `bevy-glsl-to-spirv` not supporting ` aarch64-unknown-linux-gnu`.  
I added ` aarch64-unknown-linux-gnu` to the supported platforms locally for `bevy-glsl-to-spirv` and `bevy-glsl-to-spirv` will build, but `bevy` fails on linking with various glsl components.  
  
It's not clear if Bevy 0.4 was using `shaderc` or something else in `bevy_render` broke the build.

### PR DESCRIPTION
**Pull request by mockersf**
`cfg` for `bevy-glsl-to-spirv` use now mimics https://github.com/cart/glsl-to-spirv/blob/master/Cargo.toml

fixes #898 
fixes #1348 
fixes #1942 
fixes #1078

### CODE REVIEW

File: crates/bevy_render/src/shader/shader.rs [hunk 1]
```diff
@@ -27,24 +27,59 @@ pub enum ShaderError {
...
+    #[cfg(not(any(
+        target_arch = "wasm32",
+        all(target_arch = "x86_64", target_os = "linux", target_env = "gnu"),
+        all(target_arch = "x86_64", target_os = "macos"),
+        all(target_arch = "aarch64", target_os = "android"),
+        all(target_arch = "armv7", target_os = "android"),
+        all(target_arch = "x86_64", target_os = "windows", target_env = "msvc"),
+        all(target_arch = "x86", target_os = "windows", target_env = "msvc"),
+        all(target_arch = "x86_64", target_os = "windows", target_env = "gnu"),
+        all(target_arch = "x86", target_os = "windows", target_env = "gnu")
```
Comment (MinerSebas) [inline crates/bevy_render/src/shader/shader.rs:39]: The Version of `glsl-to-spirv` on crates.io, does not yet have the Support for these three Targets.
Either they should be removed for now, or Cart should release a new version of `glsl-to-spirv`.

Comment (cart) [reply crates/bevy_render/src/shader/shader.rs:39]: I just tried publishing the new version, but the new crate structure causes problems: https://github.com/cart/glsl-to-spirv/issues/12.

Short term if we want to merge this asap we should assume we need to use the current version here.

Comment (mockersf) [reply crates/bevy_render/src/shader/shader.rs]: I removed them

Comment (mockersf) [reply crates/bevy_render/src/shader/shader.rs]: >  if we want to merge this
I opened this after seeing several people with issue around that, but none in the last few days.

**Comment by MinerSebas** [comment]
Could you edit the PR Description to include `Fixes #1348` and `Fixes #898` so that these #1348 and #898 are automatically closed when this is merged.

File: crates/bevy_render/src/shader/shader.rs [hunk 2]
```diff
@@ -27,24 +27,47 @@ pub enum ShaderError {
     #[error("Shader compilation error:\n{0}")]
     Compilation(String),
 
-    #[cfg(any(target_os = "ios", all(target_arch = "aarch64", target_os = "macos")))]
+    #[cfg(not(any(
+        target_arch = "wasm32",
+        all(target_arch = "x86_64", target_os = "linux", target_env = "gnu"),
+        all(target_arch = "x86_64", target_os = "macos"),
+        all(target_arch = "aarch64", target_os = "android"),
+        all(target_arch = "armv7", target_os = "android"),
```
Comment (MinerSebas) [inline crates/bevy_render/src/shader/shader.rs:35]:
```suggestion
        all(target_arch = "armv7", target_os = "androidabi"),
```
This must be used, to fix the CI

File: crates/bevy_render/src/shader/shader.rs [hunk 3]
```diff
@@ -27,24 +27,47 @@ pub enum ShaderError {
...
     #[error("shaderc error: {}")]
     ShaderC(#[from] shaderc::Error),
 
-    #[cfg(any(target_os = "ios", all(target_arch = "aarch64", target_os = "macos")))]
+    #[cfg(not(any(
+        target_arch = "wasm32",
+        all(target_arch = "x86_64", target_os = "linux", target_env = "gnu"),
+        all(target_arch = "x86_64", target_os = "macos"),
+        all(target_arch = "aarch64", target_os = "android"),
+        all(target_arch = "armv7", target_os = "android"),
```
Comment (MinerSebas) [inline crates/bevy_render/src/shader/shader.rs:47]:
```suggestion
        all(target_arch = "armv7", target_os = "androidabi"),
```

File: crates/bevy_render/src/shader/shader.rs [hunk 4]
```diff
@@ -27,24 +27,47 @@ pub enum ShaderError {
...
     #[error("Error initializing shaderc Compiler")]
     ErrorInitializingShadercCompiler,
 
-    #[cfg(any(target_os = "ios", all(target_arch = "aarch64", target_os = "macos")))]
+    #[cfg(not(any(
+        target_arch = "wasm32",
+        all(target_arch = "x86_64", target_os = "linux", target_env = "gnu"),
+        all(target_arch = "x86_64", target_os = "macos"),
+        all(target_arch = "aarch64", target_os = "android"),
+        all(target_arch = "armv7", target_os = "android"),
```
Comment (MinerSebas) [inline crates/bevy_render/src/shader/shader.rs:58]:
```suggestion
        all(target_arch = "armv7", target_os = "androidabi"),
```

File: crates/bevy_render/src/shader/shader.rs [hunk 5]
```diff
@@ -27,24 +27,47 @@ pub enum ShaderError {
...
 
-#[cfg(all(
-    not(target_os = "ios"),
-    not(target_arch = "wasm32"),
-    not(all(target_arch = "aarch64", target_os = "macos"))
+#[cfg(any(
+    all(target_arch = "x86_64", target_os = "linux", target_env = "gnu"),
+    all(target_arch = "x86_64", target_os = "macos"),
+    all(target_arch = "aarch64", target_os = "android"),
+    all(target_arch = "armv7", target_os = "android"),
```
Comment (MinerSebas) [inline crates/bevy_render/src/shader/shader.rs:69]:
```suggestion
    all(target_arch = "armv7", target_os = "androidabi"),
```

File: crates/bevy_render/src/shader/shader.rs [hunk 6]
```diff
@@ -56,10 +79,12 @@ impl From<ShaderStage> for bevy_glsl_to_spirv::ShaderType {
...
```
Comment (MinerSebas) [inline crates/bevy_render/src/shader/shader.rs:86]:
```suggestion
    all(target_arch = "armv7", target_os = "androidabi"),
```

File: crates/bevy_render/src/shader/shader.rs [hunk 7]
```diff
@@ -70,7 +95,14 @@ pub fn glsl_to_spirv(
         .map_err(ShaderError::Compilation)
 }
 
-#[cfg(any(target_os = "ios", all(target_arch = "aarch64", target_os = "macos")))]
+#[cfg(not(any(
+    target_arch = "wasm32",
+    all(target_arch = "x86_64", target_os = "linux", target_env = "gnu"),
+    all(target_arch = "x86_64", target_os = "macos"),
+    all(target_arch = "aarch64", target_os = "android"),
+    all(target_arch = "armv7", target_os = "android"),
```
Comment (MinerSebas) [inline crates/bevy_render/src/shader/shader.rs:103]:
```suggestion
    all(target_arch = "armv7", target_os = "androidabi"),
```

File: crates/bevy_render/src/shader/shader.rs [hunk 8]
```diff
@@ -81,7 +113,14 @@ impl Into<shaderc::ShaderKind> for ShaderStage {
     }
 }
 
-#[cfg(any(target_os = "ios", all(target_arch = "aarch64", target_os = "macos")))]
+#[cfg(not(any(
+    target_arch = "wasm32",
+    all(target_arch = "x86_64", target_os = "linux", target_env = "gnu"),
+    all(target_arch = "x86_64", target_os = "macos"),
+    all(target_arch = "aarch64", target_os = "android"),
+    all(target_arch = "armv7", target_os = "android"),
```
Comment (MinerSebas) [inline crates/bevy_render/src/shader/shader.rs:121]:
```suggestion
    all(target_arch = "armv7", target_os = "androidabi"),
```

File: crates/bevy_render/Cargo.toml [hunk 9]
```diff
@@ -44,10 +44,10 @@ parking_lot = "0.11.0"
...
+[target.'cfg(any(all(target_arch="x86_64", target_os="linux", target_env="gnu"), all(target_arch="x86_64", target_os="macos"), all(target_arch="aarch64", target_os="android"), all(target_arch="armv7", target_os="android"), all(target_arch="x86_64", target_os="windows", target_env="msvc"), all(target_arch="x86", target_os="windows", target_env="msvc"), all(target_arch="x86_64", target_os="windows", target_env="gnu"), all(target_arch="x86", target_os="windows", target_env="gnu")))'.dependencies]
 bevy-glsl-to-spirv = "0.2.0"
 
-[target.'cfg(any(target_os = "ios", all(target_arch = "aarch64", target_os = "macos")))'.dependencies]
+[target.'cfg(not(any(target_arch = "wasm32", all(target_arch="x86_64", target_os="linux", target_env="gnu"), all(target_arch="x86_64", target_os="macos"), all(target_arch="aarch64", target_os="android"), all(target_arch="armv7", target_os="android"), all(target_arch="x86_64", target_os="windows", target_env="msvc"), all(target_arch="x86", target_os="windows", target_env="msvc"), all(target_arch="x86_64", target_os="windows", target_env="gnu"), all(target_arch="x86", target_os="windows", target_env="gnu"))))'.dependencies]
```
Comment (MinerSebas) [inline crates/bevy_render/Cargo.toml:50]:
```suggestion
[target.'cfg(not(any(target_arch = "wasm32", all(target_arch="x86_64", target_os="linux", target_env="gnu"), all(target_arch="x86_64", target_os="macos"), all(target_arch="aarch64", target_os="android"), all(target_arch="armv7", target_os="androidabi"), all(target_arch="x86_64", target_os="windows", target_env="msvc"), all(target_arch="x86", target_os="windows", target_env="msvc"), all(target_arch="x86_64", target_os="windows", target_env="gnu"), all(target_arch="x86", target_os="windows", target_env="gnu"))))'.dependencies]
```

**Comment by skhameneh** [comment]
Confirming https://github.com/bevyengine/bevy/commit/a46c35a7a050440fa3e672d46bad97c94ed8a0a9 fixes https://github.com/bevyengine/bevy/issues/1942.

File: crates/bevy_render/Cargo.toml [hunk 10]
```diff
@@ -44,10 +44,10 @@ parking_lot = "0.11.0"
 [target.'cfg(not(target_arch = "wasm32"))'.dependencies]
 spirv-reflect = "0.2.3"
 
-[target.'cfg(all(not(target_os = "ios"), not(target_arch = "wasm32"), not(all(target_arch = "aarch64", target_os = "macos"))))'.dependencies]
+[target.'cfg(any(all(target_arch="x86_64", target_os="linux", target_env="gnu"), all(target_arch="x86_64", target_os="macos"), all(target_arch="aarch64", target_os="android"), all(target_arch="armv7", target_os="androidabi"), all(target_arch="x86_64", target_os="windows", target_env="msvc"), all(target_arch="x86", target_os="windows", target_env="msvc"), all(target_arch="x86_64", target_os="windows", target_env="gnu"), all(target_arch="x86", target_os="windows", target_env="gnu")))'.dependencies]
```
Comment (MinerSebas) [inline crates/bevy_render/Cargo.toml:47]:
```suggestion
[target.'cfg(any(all(target_arch="x86_64", target_os="linux", target_env="gnu"), all(target_arch="x86_64", target_os="macos"), all(target_arch="aarch64", target_os="androidabi"), all(target_arch="armv7", target_os="androidabi"), all(target_arch="x86_64", target_os="windows", target_env="msvc")))'.dependencies]
```

File: crates/bevy_render/Cargo.toml [hunk 11]
```diff
@@ -44,10 +44,10 @@ parking_lot = "0.11.0"
...
 bevy-glsl-to-spirv = "0.2.0"
 
-[target.'cfg(any(target_os = "ios", all(target_arch = "aarch64", target_os = "macos")))'.dependencies]
+[target.'cfg(not(any(target_arch = "wasm32", all(target_arch="x86_64", target_os="linux", target_env="gnu"), all(target_arch="x86_64", target_os="macos"), all(target_arch="aarch64", target_os="android"), all(target_arch="armv7", target_os="androidabi"), all(target_arch="x86_64", target_os="windows", target_env="msvc"), all(target_arch="x86", target_os="windows", target_env="msvc"), all(target_arch="x86_64", target_os="windows", target_env="gnu"), all(target_arch="x86", target_os="windows", target_env="gnu"))))'.dependencies]
```
Comment (MinerSebas) [inline crates/bevy_render/Cargo.toml:50]:
```suggestion
[target.'cfg(not(any(target_arch = "wasm32", all(target_arch="x86_64", target_os="linux", target_env="gnu"), all(target_arch="x86_64", target_os="macos"), all(target_arch="aarch64", target_os="android"), all(target_arch="armv7", target_os="androidabi"), all(target_arch="x86_64", target_os="windows", target_env="msvc"))))'.dependencies]
```

**Review by MinerSebas — COMMENTED** [review]
Missed to suggest these two. :sweat_smile:

**Comment by MinerSebas** [comment]
@mockersf I found another Issue that can be closed when this merges: #1078

**Comment by cart** [comment]
This looks good to me. Not a huge fan of the conditional compilation complexity + large duplicated `target` blocks that need to be kept in sync, but this does seem like the pragmatic choice for now.

**Pull request closed without merging**